cranelift/codegen/src/isa/aarch64/inst.isle

```diff
@@ -0,0 +1,1364 @@
+    (Uaddlp16)
+))
+
+;; An operation across the lanes of vectors.
+(type VecLanesOp
+  (enum
+    ;; Integer addition across a vector
+    (Addv)
+    ;; Unsigned minimum across a vector
+    (Uminv)
+))
+
+;; A shift-by-immediate operation on each lane of a vector.
+(type VecShiftImmOp
+  (enum
+    ;; Unsigned shift left
+    (Shl)
+    ;; Unsigned shift right
+    (Ushr)
+    ;; Signed shift right
+    (Sshr)
+))
+
+;; Atomic read-modify-write operations with acquire-release semantics
+(type AtomicRMWOp
+  (enum
+    (Add)
+    (Clr)
+    (Eor)
+    (Set)
+    (Smax)
+    (Smin)
+    (Umax)
+    (Umin)
+))
+
+;; Extractor helpers for various immmediate constants ;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(decl move_wide_const_from_u64 (MoveWideConst) u64)
+(extern extractor move_wide_const_from_u64 move_wide_const_from_u64)
+
+(decl move_wide_const_from_negated_u64 (MoveWideConst) u64)
+(extern extractor move_wide_const_from_negated_u64 move_wide_const_from_negated_u64)
+
+(decl imm_logic_from_u64 (ImmLogic) u64)
+(extern extractor imm_logic_from_u64 imm_logic_from_u64)
+
+(decl integral_ty (Type) Type)
+(extern extractor integral_ty integral_ty)
+
+;; Instruction creation helpers ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+;; Emit an instruction.
+;;
+;; This is low-level and side-effectful; it should only be used as an
+;; implementation detail by helpers that preserve the SSA facade themselves.
+(decl emit (MInst) Unit)
+(extern constructor emit emit)
+
+;; Helper for creating the zero register.
+(decl zero_reg () Reg)
+(extern constructor zero_reg zero_reg)
+
+;; Helper for emitting `MInst.MovZ` instructions.
+(decl movz (MoveWideConst OperandSize) Reg)
+(rule (movz imm size)
+      (let ((dst WritableReg (temp_writable_reg $I64))
+            (_ Unit (emit (MInst.MovZ dst imm size))))
+        (writable_reg_to_reg dst)))
+
+;; Helper for emitting `MInst.MovN` instructions.
+(decl movn (MoveWideConst OperandSize) Reg)
+(rule (movn imm size)
+      (let ((dst WritableReg (temp_writable_reg $I64))
+            (_ Unit (emit (MInst.MovN dst imm size))))
+        (writable_reg_to_reg dst)))
+
+;; Helper for emitting `MInst.AluRRImmLogic` instructions.
+(decl alu_rr_imm_logic (ALUOp Reg ImmLogic) Reg)
+(rule (alu_rr_imm_logic op src imm)
+      (let ((dst WritableReg (temp_writable_reg $I64))
+            (_ Unit (emit (MInst.AluRRImmLogic op dst src imm))))
+        (writable_reg_to_reg dst)))
+
+;; Helper for emitting `orr64` instructions.
+(decl orr64 (Reg ImmLogic) Reg)
+(rule (orr64 src imm) (alu_rr_imm_logic (ALUOp.Orr64) src imm))
+
+;; Immediate value helpers ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(decl imm (Type u64) Reg)
+
+;; First up some special cases for constants which match certain custom
+;; conditions. Note that these have explicit priorities to explicitly list
+;; that we prefer instructions in the order that they're listed here. Otherwise
+;; ISLE has no way of determining which is the more relevant matcher.
```

Comment:
Someone more familiar with AArch64 may want to confirm this, I'm respecting the original ordering that's implemented today, but I'm not sure if it matters matching it precisely or if any of the three one-instruction forms is fine to materialize a constant with.

Reply:
Since they're all 1-instruction lowerings, it should be fine (and equal cost) to use any that fit, I think. (Though at least MOVZ and MOVN should be completely mutually exclusive, since all the other bits except the shifted 16-bit field are all 0 or all 1 respectively, but ORI's range overlaps with MOVZ.)

So I think it's slightly better to exclude the priorities here so the decision trie build has more freedom to merge and reorder.

Excellent attention to preserving original behavior though, thanks for taking care with that!